modules/library/main/src/main/java/org/geotools/geometry/jts/GeometryClipper.java

```diff
@@ -76,6 +79,49 @@ public GeometryClipper(Envelope bounds) {
     }
     
     /**
+     * This will try to handle failures when clipping - i.e. because of invalid input geometries (often 
+     * caused by simplification).
+     * 
+     * This attempts to do a normal clip().  If it fails, it will try to do more to ensure the clip 
+     * works properly.  
+     * 
+     * The first attempt to correct the geometry is to put the points on a precision grid and redo
+     * the clip. 
+     * If this fails, the original geometry is returned.
+     * 
+     * see {@link #clip(Geometry, boolean) clip}
+     * 
+     * @param g
+     * @param ensureValid
+     * @param scale 0=double precision Precision Model
+     * @return
+     */
+    public Geometry clipFailResistant(Geometry g, boolean ensureValid, double scale) {
```

Comment:
If you use scale=0, then it will attempt to fix invalid polygons - inside the precisionmodel code  It does a polygon.buffer(0), which will often "fix" an invalid polygon (but not always).  I can add that explicitly - its probably better to attempt the buffer(0) first, then attempt to re-compute the intersection, and then move the points around (precisionmodel).  That's likely a bit better (but more time-consuming) than that what the original code did (buffer(0) might be failing because its being applied after the precisionmodel instead of trying it both before and after).  The original code wasn't noticing the precisionmodel failures - which is what this code does.  I expect the original author thought that if the precisionmodel failed it would throw an exception.  

I could keep the original code and just add;
if (reduced.isEmpty()) {
       throw TopologyException(...);
   }

 I didn't do much work making it more robust than it was  - I didn't see any references to the clipper outside the renderer.  I'd need more test data and use cases to improve it.  The main problem is GIGO - invalid polygons aren't easy to robustly work with.  I added the function here so it could grow, instead of everyone "doing a little" in their own code (i.e. the clip() method in streamingrenderer).

I can move this function back into streamingrenderer, but seems a bit weird to write computation geometry code and test cases in the streamingrenderer instead of in the GeometryClipper (or some type subclass/wrapper).  

So, I can;
a) Move code back to StreamingRenderer, just do the .isEmpty()/throw
b) keep code where it is, but be more aggressive in trying to convert the invalid polygon to a valid polygon (i.e. add an explicit buffer(0))
c) add GeometryClipper Interface, and then a pixel-based renderer wrapper 

I'd recommend (b) or (c) - (a) is easier but I don't really think the CG code and tests should be in the rendrerer and the code is likely a useful starting place for others.

To get actual robust clipping with invalid polygons would be a very very big task (and we havent even talked about what to do when JTS cannot compute the intersection even with valid polygons).  Everyone might have a different opinion on what to do - and likely require some type of hints into the function (i.e. perturbPoints=allowed, moveClipRectange=allowed, etc...).  Like I said - need lots of use cases and test cases.  For example, it might make sense to, if the polygon is invalid, use the simpler non-robust clipping method since the input is already invalid and likely ok to return something invalid.